### PR TITLE
add utils.set_determinism for reproducibility

### PR DESCRIPTION
**Pull request by lessw2020**
This PR:
1 - adds a set_determinism function to set seed for Python, PyTorch, CUDA and deterministic settings for cudnn.
2 - if seed ==1, then no deterministic settings are used.  This may be important as turning off cuDnn benchmarking to ensure determinism, can also negatively impact perf.

This lack of determinism negatively impacted work with AWS where we ended up with variations in loss curves while running fp8 for our joint blog that appeared to be from fp8 but are instead likely due to not having determinism in titan. 

Testing - I ran multiple small runs with 7B while rotating between three seeds and saw consistent ending loss points matching to the respective seeds.  

This PR does not set per worker aspects for the dataloader since we do not shuffle atm...but that could become a future source of randomness that will need to be set if we shuffle in the future.

### CODE REVIEW

File: torchtitan/config_manager.py [hunk 1]
```diff
@@ -357,7 +357,12 @@ def __init__(self):
             default=50,
             help="Python garbage control scheduling interval, in steps",
         )
-
+        self.parser.add_argument(
+            "--training.seed",
+            type=int,
+            default=2024,
```
Comment (tianyu-l) [inline torchtitan/config_manager.py]: Let's set default to `None`

File: torchtitan/utils.py [hunk 2]
```diff
@@ -15,6 +15,7 @@
 import torch.distributed._functional_collectives as funcol
 import torch.distributed.distributed_c10d as c10d
 from torch.distributed.device_mesh import DeviceMesh
+from torchtitan.config_manager import JobConfig
```
Comment (tianyu-l) [inline torchtitan/utils.py]: don't need this import

File: torchtitan/utils.py [hunk 3]
```diff
@@ -36,6 +37,24 @@ def _warn_overwrite_env(env, val):
     os.environ[env] = val
 
 
+def set_determinism(job_config: JobConfig) -> None:
```
Comment (tianyu-l) [inline torchtitan/utils.py]: just a `random_seed` as input arg, of type `Optional[int]`

File: torchtitan/utils.py [hunk 4]
```diff
@@ -36,6 +37,24 @@ def _warn_overwrite_env(env, val):
...
+    """
+    Set Python, PyTorch, CUDA seeds and cudnn settings for reproducibility
+    Use 1 to avoid determinism
+    """
+    if job_config.training.seed == 1:
```
Comment (tianyu-l) [inline torchtitan/utils.py]: let's use `None` instead of `1` to disable determinism, as `1` is a valid seed

File: torchtitan/utils.py [hunk 5]
```diff
@@ -36,6 +37,24 @@ def _warn_overwrite_env(env, val):
     os.environ[env] = val
 
 
+def set_determinism(job_config: JobConfig) -> None:
+    """
+    Set Python, PyTorch, CUDA seeds and cudnn settings for reproducibility
+    Use 1 to avoid determinism
```
Comment (tianyu-l) [inline torchtitan/utils.py]: pls change comment as well

File: train.py [hunk 6]
```diff
@@ -56,6 +56,10 @@ def main(job_config: JobConfig):
     # take control of garbage collection to avoid stragglers
     gc_handler = utils.GarbageCollection(gc_freq=job_config.training.gc_freq)
 
+    # set determinisism, use seed == 1 to skip random seed setting
+    utils.set_determinism(job_config)
+    logger.info(f"Using seed: {job_config.training.seed}")
```
Comment (tianyu-l) [inline train.py]: we can say "Deterministic training off" or "Deterministic training on with seed xxx" depending on whether `seed is None`

File: train_configs/debug_model.toml [hunk 7]
```diff
@@ -30,6 +30,7 @@ name = "AdamW"
 lr = 8e-4
 
 [training]
+seed = 1  # use 1 avoid determinism
```
Comment (tianyu-l) [inline train_configs/debug_model.toml]: Since the default value is `None` we can remove this flag in toml file.

**Review by tianyu-l — COMMENTED** [review]
Agree that this is a missing functionality we should have!
Left inline comments on how I think we can better expose it.

File: torchtitan/utils.py [hunk 8]
```diff
@@ -36,6 +37,24 @@ def _warn_overwrite_env(env, val):
...
+    """
+    if job_config.training.seed == 1:
+        return
+    # CPU determinism
+    torch.manual_seed(job_config.training.seed)
+    # GPU determinism
+    torch.cuda.manual_seed_all(job_config.training.seed)
+    # set deterministic cudnn algorithms
+    torch.backends.cudnn.deterministic = True
```
Comment (fegin) [inline torchtitan/utils.py]: Will it be better to call `torch.use_deterministic_algorithms(True)`?

Comment (lessw2020) [reply torchtitan/utils.py]: actually i tried that earlier, and it causes torch.compile to break...so I did not use it.

Comment (fegin) [reply torchtitan/utils.py]: According to https://pytorch.org/docs/stable/notes/randomness.html#cuda-convolution-determinism, this seems to be the right way if we want everything to be deterministic. Should we raise an issue to the compiler team?

Comment (lessw2020) [reply torchtitan/utils.py]: oh ok, thanks @fegin  - yes I will repro this tomorrow and log the issue with them.

File: train_configs/debug_model.toml (hunk 7 above)
Comment (lessw2020) [reply train_configs/debug_model.toml]: removed

File: train.py (hunk 6 above)
Comment (lessw2020) [reply train.py]: yes, added.

File: torchtitan/utils.py (hunk 5 above)
Comment (lessw2020) [reply torchtitan/utils.py]: udpdated.

File: torchtitan/utils.py (hunk 4 above)
Comment (lessw2020) [reply torchtitan/utils.py]: sounds good.  I went with 1 b/c we use 1 as off for TP etc. and b/c None is not valid in yaml...but None is better and we just avoid having it in toml by making it default in config_manager.py.

File: torchtitan/utils.py (hunk 3 above)
Comment (lessw2020) [reply torchtitan/utils.py]: updated, though just used 'seed' because putting random with it when it's used for determinism makes it potentially confusing imo.

File: torchtitan/utils.py (hunk 2 above)
Comment (lessw2020) [reply torchtitan/utils.py]: removed

File: torchtitan/config_manager.py (hunk 1 above)
Comment (lessw2020) [reply torchtitan/config_manager.py]: updated.

File: torchtitan/utils.py [hunk 9]
```diff
@@ -36,6 +36,26 @@ def _warn_overwrite_env(env, val):
     os.environ[env] = val
 
 
+def set_determinism(seed: Optional[int]) -> None:
+    """
+    Set Python, PyTorch, CUDA seeds and cudnn settings for reproducibility
+    """
+    if seed is not None:
+        # CPU determinism
+        torch.manual_seed(seed)
```
Comment (awgu) [inline torchtitan/utils.py:45]: Seems this sets the seed for all devices, not just CPU: https://pytorch.org/docs/stable/generated/torch.manual_seed.html

Comment (tianyu-l) [reply torchtitan/utils.py:45]: Good to know! Yes let's remove redundant `torch.cuda.manual_seed_all(seed)` here.

Comment (lessw2020) [reply torchtitan/utils.py:45]: removed

**Review by tianyu-l — APPROVED** [review]
LGTM! Thanks!

**Comment by lessw2020** [comment]
4 gpu failure is unrelated to this PR (issue with chown on a file).

File: torchtitan/utils.py (hunk 8 above)
Comment (lessw2020) [reply torchtitan/utils.py]: I filed the promised issue:
https://github.com/pytorch/pytorch/issues/137160
